File: test/api_smoke_test.py
#!/usr/bin/env python3
"""API smoke test for the Rust Todo backend.

Usage examples:
  python3 test/api_smoke_test.py
  python3 test/api_smoke_test.py -d
  python3 test/api_smoke_test.py --base-url http://127.0.0.1:5050
  python3 test/api_smoke_test.py --start-server

Notes:
- Requires MongoDB at mongodb://127.0.0.1:27017
- If --start-server is not used, the backend must already be running.
"""

from __future__ import annotations

import argparse
import json
import random
import string
import subprocess
import sys
import time
from pathlib import Path
from typing import Any, Optional, Tuple
from urllib import error, parse, request


DEFAULT_BASE_URL = "http://127.0.0.1:5050"
USE_COLOR = sys.stdout.isatty()


class Ansi:
    RESET = "\033[0m"
    CYAN = "\033[36m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    RED = "\033[31m"


def _colorize(text: str, color: str, enabled: bool) -> str:
    if not enabled:
        return text
    return f"{color}{text}{Ansi.RESET}"


class TestFailure(Exception):
    pass


class ApiClient:
    def __init__(
        self,
        base_url: str,
        timeout_seconds: float = 10.0,
        debug: bool = False,
        color: bool = True,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout_seconds = timeout_seconds
        self.debug = debug
        self.color = color

    def request_json(
        self,
        method: str,
        path: str,
        payload: Optional[dict[str, Any]] = None,
        query: Optional[dict[str, Any]] = None,
        trace: bool = True,
    ) -> Tuple[int, Any]:
        method = method.upper()
        url = f"{self.base_url}{path}"
        if query:
            url = f"{url}?{parse.urlencode(query)}"

        headers = {
            "Accept": "application/json",
            "Content-Type": "application/json",
        }
        data = None
        if payload is not None:
            data = json.dumps(payload).encode("utf-8")

        if self.debug and trace:
            self._print_debug_request(method, url, payload)

        req = request.Request(url=url, data=data, headers=headers, method=method)

        try:
            with request.urlopen(req, timeout=self.timeout_seconds) as resp:
                raw = resp.read().decode("utf-8")
                body = _parse_json_or_text(raw)
                if self.debug and trace:
                    self._print_debug_response(resp.status, body)
                return resp.status, body
        except error.HTTPError as exc:
            raw = exc.read().decode("utf-8")
            body = _parse_json_or_text(raw)
            if self.debug and trace:
                self._print_debug_response(exc.code, body)
            return exc.code, body

    def _print_debug_request(
        self,
        method: str,
        url: str,
        payload: Optional[dict[str, Any]],
    ) -> None:
        label = _colorize("[DEBUG REQUEST]", Ansi.CYAN, self.color)
        print(f"{label} {method} {url}")
        print("Request JSON:")
        print(_pretty_json(payload))

    def _print_debug_response(self, status: int, body: Any) -> None:
        response_color = Ansi.GREEN if status < 400 else Ansi.RED
        label = _colorize("[DEBUG RESPONSE]", response_color, self.color)
        print(f"{label} HTTP {status}")
        print("Response JSON:")
        print(_pretty_json(body))
        print()


def _parse_json_or_text(raw: str) -> Any:
    if not raw:
        return None
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return raw


def _pretty_json(value: Any) -> str:
    try:
        return json.dumps(value, indent=2, sort_keys=True)
    except TypeError:
        return str(value)


def _wait_for_server(
    client: ApiClient,
    timeout_seconds: int = 45,
    server_proc: Optional[subprocess.Popen[str]] = None,
) -> None:
    deadline = time.time() + timeout_seconds
    while time.time() < deadline:
        if server_proc is not None and server_proc.poll() is not None:
            output = ""
            if server_proc.stdout is not None:
                output = server_proc.stdout.read().strip()
            details = f"\nServer output:\n{output}" if output else ""
            raise TestFailure(
                "Backend process exited before becoming ready."
                + details
            )
        try:
            status, _ = client.request_json(
                "GET",
                "/get-user",
                query={"email": "_probe_"},
                trace=False,
            )
            if status == 200:
                return
        except Exception:
            pass
        time.sleep(1)
    raise TestFailure(
        f"Server at {client.base_url} was not reachable within {timeout_seconds}s"
    )


def _start_server(project_root: Path) -> subprocess.Popen[str]:
    cmd = ["cargo", "run", "--quiet"]
    proc = subprocess.Popen(
        cmd,
        cwd=project_root,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        bufsize=1,
    )
    return proc


def _stop_server(proc: subprocess.Popen[str]) -> None:
    if proc.poll() is not None:
        return

    proc.terminate()
    try:
        proc.wait(timeout=8)
    except subprocess.TimeoutExpired:
        proc.kill()
        proc.wait(timeout=5)


def _assert(condition: bool, message: str) -> None:
    if not condition:
        raise TestFailure(message)


def _step(name: str) -> None:
    print(f"{_colorize('[STEP]', Ansi.YELLOW, USE_COLOR)} {name}")


def _ok(name: str) -> None:
    print(f"{_colorize('[PASS]', Ansi.GREEN, USE_COLOR)} {name}")


def _get_task_by_id(tasks: list[dict[str, Any]], task_id: int) -> dict[str, Any]:
    for task in tasks:
        if task.get("id") == task_id:
            return task
    raise TestFailure(f"Task id={task_id} not found in tasks payload: {tasks}")


def run_smoke_test(client: ApiClient) -> None:
    suffix = f"{int(time.time())}_{_rand(6)}"
    username = f"test_user_{suffix}"
    email = f"todo.{suffix}@example.com"
    password = f"Pass_{suffix}!"

    _step("Create a new user")
    status, body = client.request_json(
        "POST",
        "/create-user",
        {
            "username": username,
            "email": email,
            "password": password,
        },
    )
    _assert(status == 200, f"/create-user expected HTTP 200, got {status}, body={body}")
    _assert(isinstance(body, dict), f"/create-user expected JSON object, got {body}")
    _assert(body.get("status") == "Success", f"/create-user failed: body={body}")
    _ok("User created")

    _step("Fetch user by email")
    status, body = client.request_json("GET", "/get-user", query={"email": email})
    _assert(status == 200, f"/get-user expected HTTP 200, got {status}, body={body}")
    _assert(body.get("email") == email, f"/get-user email mismatch: body={body}")
    _assert(body.get("username") == username, f"/get-user username mismatch: body={body}")
    _ok("User fetched")

    _step("Authenticate with wrong password (negative test)")
    status, body = client.request_json(
        "POST",
        "/auth",
        {
            "email": email,
            "pass": f"{password}_wrong",
        },
    )
    _assert(status == 200, f"/auth (wrong pass) expected HTTP 200, got {status}, body={body}")
    _assert(body.get("auth") is False, f"/auth (wrong pass) auth should be false: body={body}")
    _assert(body.get("error") is True, f"/auth (wrong pass) error should be true: body={body}")
    _assert(body.get("jwt") is None, f"/auth (wrong pass) jwt should be null: body={body}")
    _ok("Wrong password rejected")

    _step("Authenticate with valid password")
    status, body = client.request_json(
        "POST",
        "/auth",
        {
            "email": email,
            "pass": password,
        },
    )
    _assert(status == 200, f"/auth expected HTTP 200, got {status}, body={body}")
    _assert(body.get("auth") is True, f"/auth should return auth=true: body={body}")
    _assert(body.get("error") is False, f"/auth should return error=false: body={body}")
    jwt = body.get("jwt")
    _assert(isinstance(jwt, str) and len(jwt) > 10, f"/auth jwt missing/invalid: body={body}")
    _ok("Authentication succeeded")

    _step("Verify valid JWT")
    status, body = client.request_json("POST", "/jwt-auth", {"jwt": jwt})
    _assert(status == 200, f"/jwt-auth expected HTTP 200, got {status}, body={body}")
    _assert(body.get("tampered") is False, f"/jwt-auth tampered should be false: body={body}")
    _assert(body.get("email") == email, f"/jwt-auth email mismatch: body={body}")
    _ok("Valid JWT accepted")

    _step("Verify tampered JWT (negative test)")
    status, body = client.request_json("POST", "/jwt-auth", {"jwt": f"{jwt}.broken"})
    _assert(status == 200, f"/jwt-auth (tampered) expected HTTP 200, got {status}, body={body}")
    _assert(body.get("tampered") is True, f"/jwt-auth (tampered) should be true: body={body}")
    _assert(body.get("email") is None, f"/jwt-auth (tampered) email should be null: body={body}")
    _ok("Tampered JWT rejected")

    _step("Add two tasks")
    for idx, name in enumerate(("Write tests", "Ship release"), start=1):
        status, body = client.request_json(
            "POST",
            "/add-task",
            {
                "name": name,
                "priority": idx,
                "auth_jwt": jwt,
            },
        )
        _assert(status == 200, f"/add-task expected HTTP 200, got {status}, body={body}")
        _assert(body is True, f"/add-task expected true, got body={body}")
    _ok("Tasks added")

    _step("Get tasks with authenticated request")
    status, body = client.request_json(
        "GET",
        "/get-tasks",
        {
            "auth_jwt": jwt,
            "counter": 10,
        },
    )
    _assert(status == 200, f"/get-tasks expected HTTP 200, got {status}, body={body}")
    _assert(body.get("res") is True, f"/get-tasks should return res=true: body={body}")
    tasks = body.get("tasks")
    _assert(isinstance(tasks, list), f"/get-tasks tasks should be a list: body={body}")
    _assert(len(tasks) >= 1, f"/get-tasks expected at least one task, got {body}")
    first_task = tasks[0]
    _assert("id" in first_task and "name" in first_task, f"Unexpected task shape: {first_task}")
    _ok("Tasks fetched")

    first_task_id = first_task["id"]
    initial_status = first_task.get("status")
    _assert(isinstance(initial_status, bool), f"Task status should be bool: {first_task}")

    _step("Toggle first task status by id")
    status, body = client.request_json(
        "POST",
        "/update-task",
        {
            "id": first_task_id,
            "auth_jwt": jwt,
        },
    )
    _assert(status == 200, f"/update-task expected HTTP 200, got {status}, body={body}")
    _assert(body.get("success") is True, f"/update-task expected success=true, got {body}")
    _ok("Task status toggled once")

    _step("Verify first toggle changed task status")
    status, body = client.request_json(
        "GET",
        "/get-tasks",
        {
            "auth_jwt": jwt,
            "counter": 10,
        },
    )
    _assert(status == 200, f"/get-tasks (after 1st toggle) expected HTTP 200, got {status}, body={body}")
    _assert(body.get("res") is True, f"/get-tasks (after 1st toggle) should return res=true: body={body}")
    tasks = body.get("tasks")
    _assert(isinstance(tasks, list), f"/get-tasks (after 1st toggle) tasks should be list: body={body}")
    toggled_task = _get_task_by_id(tasks, first_task_id)
    _assert(
        toggled_task.get("status") is (not initial_status),
        f"/update-task should toggle status from {initial_status} to {not initial_status}, got {toggled_task}",
    )
    _ok("First toggle verified")

    _step("Toggle first task status back to original value")
    status, body = client.request_json(
        "POST",
        "/update-task",
        {
            "id": first_task_id,
            "auth_jwt": jwt,
        },
    )
    _assert(status == 200, f"/update-task (2nd toggle) expected HTTP 200, got {status}, body={body}")
    _assert(body.get("success") is True, f"/update-task (2nd toggle) expected success=true, got {body}")
    _ok("Task status toggled twice")

    _step("Verify second toggle restored original task status")
    status, body = client.request_json(
        "GET",
        "/get-tasks",
        {
            "auth_jwt": jwt,
            "counter": 10,
        },
    )
    _assert(status == 200, f"/get-tasks (after 2nd toggle) expected HTTP 200, got {status}, body={body}")
    _assert(body.get("res") is True, f"/get-tasks (after 2nd toggle) should return res=true: body={body}")
    tasks = body.get("tasks")
    _assert(isinstance(tasks, list), f"/get-tasks (after 2nd toggle) tasks should be list: body={body}")
    toggled_back_task = _get_task_by_id(tasks, first_task_id)
    _assert(
        toggled_back_task.get("status") is initial_status,
        f"/update-task second toggle should restore status={initial_status}, got {toggled_back_task}",
    )
    _ok("Second toggle verified")

    _step("Update unknown task id (negative test)")
    status, body = client.request_json(
        "POST",
        "/update-task",
        {
            "id": first_task_id + 10_000_000,
            "auth_jwt": jwt,
        },
    )
    _assert(
        status == 500,
        f"/update-task (unknown task id) expected HTTP 500, got {status}, body={body}",
    )
    _assert(
        body.get("success") is False,
        f"/update-task (unknown task id) should return success=false, got {body}",
    )
    _ok("Unknown task id rejected")

    _step("Delete task with invalid JWT (negative test)")
    status, body = client.request_json(
        "POST",
        "/delete-task",
        {
            "auth_jwt": f"{jwt}.broken",
            "task_id": first_task_id,
        },
    )
    _assert(status == 500, f"/delete-task (invalid jwt) expected HTTP 500, got {status}, body={body}")
    _assert(body.get("success") is False, f"/delete-task (invalid jwt) should fail: body={body}")
    _ok("Invalid-JWT delete rejected")

    _step("Delete task with valid JWT")
    status, body = client.request_json(
        "POST",
        "/delete-task",
        {
            "auth_jwt": jwt,
            "task_id": first_task_id,
        },
    )
    _assert(
        status == 200,
        f"/delete-task (correct pass) expected HTTP 200, got {status}, body={body}",
    )
    _assert(body.get("success") is True, f"/delete-task should return success=true: body={body}")
    _ok("Task deleted")


def _rand(length: int) -> str:
    alphabet = string.ascii_lowercase + string.digits
    return "".join(random.choice(alphabet) for _ in range(length))


def main() -> int:
    parser = argparse.ArgumentParser(description="Smoke test the Todo backend API")
    parser.add_argument("--base-url", default=DEFAULT_BASE_URL, help="API base URL")
    parser.add_argument(
        "-d",
        "--debug",
        action="store_true",
        help="Print colorized request/response JSON for each API call",
    )
    parser.add_argument(
        "--start-server",
        action="store_true",
        help="Start backend with `cargo run --quiet` before tests",
    )
    args = parser.parse_args()

    client = ApiClient(args.base_url, debug=args.debug, color=USE_COLOR)
    server_proc: Optional[subprocess.Popen[str]] = None

    try:
        if args.start_server:
            project_root = Path(__file__).resolve().parent.parent
            print("[INFO] Starting backend server with cargo run --quiet")
            server_proc = _start_server(project_root)

        print(f"[INFO] Waiting for server: {args.base_url}")
        _wait_for_server(client, server_proc=server_proc)

        print("[INFO] Running smoke test flow")
        run_smoke_test(client)

        print("\n[SUCCESS] All smoke tests passed")
        return 0
    except TestFailure as exc:
        print(f"\n[FAIL] {exc}")
        return 1
    except Exception as exc:
        print(f"\n[ERROR] Unexpected exception: {exc}")
        return 1
    finally:
        if server_proc is not None:
            print("[INFO] Stopping backend server")
            _stop_server(server_proc)


if __name__ == "__main__":
    sys.exit(main())

File: test/api_smoke_test.js
#!/usr/bin/env bun

/**
 * API smoke test for the Rust Todo backend.
 *
 * Usage examples:
 *   bun test/api_smoke_test.js
 *   bun test/api_smoke_test.js -d
 *   bun test/api_smoke_test.js --base-url http://127.0.0.1:5050
 *   bun test/api_smoke_test.js --start-server
 *
 * Notes:
 * - Requires MongoDB at mongodb://127.0.0.1:27017
 * - If --start-server is not used, the backend must already be running.
 */

import { spawn } from "node:child_process";
import http from "node:http";
import https from "node:https";
import path from "node:path";
import process from "node:process";
import { setTimeout as sleep } from "node:timers/promises";
import { fileURLToPath } from "node:url";

const DEFAULT_BASE_URL = "http://127.0.0.1:5050";
const USE_COLOR = Boolean(process.stdout.isTTY);

const Ansi = {
  RESET: "\x1b[0m",
  CYAN: "\x1b[36m",
  GREEN: "\x1b[32m",
  YELLOW: "\x1b[33m",
  RED: "\x1b[31m",
};

class TestFailure extends Error {}

function colorize(text, color, enabled) {
  return enabled ? `${color}${text}${Ansi.RESET}` : text;
}

function parseJsonOrText(raw) {
  if (!raw) {
    return null;
  }

  try {
    return JSON.parse(raw);
  } catch {
    return raw;
  }
}

function prettyJson(value) {
  try {
    return JSON.stringify(value, null, 2);
  } catch {
    return String(value);
  }
}

function assert(condition, message) {
  if (!condition) {
    throw new TestFailure(message);
  }
}

function step(name) {
  console.log(`${colorize("[STEP]", Ansi.YELLOW, USE_COLOR)} ${name}`);
}

function ok(name) {
  console.log(`${colorize("[PASS]", Ansi.GREEN, USE_COLOR)} ${name}`);
}

function rand(length) {
  const alphabet = "abcdefghijklmnopqrstuvwxyz0123456789";
  let out = "";
  for (let i = 0; i < length; i += 1) {
    out += alphabet[Math.floor(Math.random() * alphabet.length)];
  }
  return out;
}

function getTaskById(tasks, taskId) {
  const task = tasks.find((t) => t?.id === taskId);
  if (!task) {
    throw new TestFailure(`Task id=${taskId} not found in tasks payload: ${prettyJson(tasks)}`);
  }
  return task;
}

function getProjectRoot() {
  const scriptDir = path.dirname(fileURLToPath(import.meta.url));
  return path.resolve(scriptDir, "..");
}

class ApiClient {
  constructor(baseUrl, timeoutMs = 10_000, debug = false, color = true) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.timeoutMs = timeoutMs;
    this.debug = debug;
    this.color = color;
  }

  async requestJson(method, path, payload = null, query = null, trace = true) {
    const upperMethod = method.toUpperCase();
    const url = new URL(`${this.baseUrl}${path}`);

    if (query) {
      for (const [key, value] of Object.entries(query)) {
        url.searchParams.set(key, String(value));
      }
    }

    if (this.debug && trace) {
      this.printDebugRequest(upperMethod, url.toString(), payload);
    }

    // fetch() does not allow GET requests with a body, but this API's
    // /get-tasks endpoint expects JSON payload on GET.
    if (upperMethod === "GET" && payload !== null) {
      const [status, body] = await this.requestViaNodeHttp(upperMethod, url, payload);
      if (this.debug && trace) {
        this.printDebugResponse(status, body);
      }
      return [status, body];
    }

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const response = await fetch(url, {
        method: upperMethod,
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: payload === null ? undefined : JSON.stringify(payload),
        signal: controller.signal,
      });

      const raw = await response.text();
      const body = parseJsonOrText(raw);

      if (this.debug && trace) {
        this.printDebugResponse(response.status, body);
      }

      return [response.status, body];
    } catch (error) {
      if (error instanceof Error && error.name === "AbortError") {
        throw new TestFailure(`Request timed out after ${this.timeoutMs}ms: ${upperMethod} ${url}`);
      }
      throw error;
    } finally {
      clearTimeout(timeout);
    }
  }

  requestViaNodeHttp(method, url, payload) {
    const transport = url.protocol === "https:" ? https : http;
    const bodyString = payload === null ? "" : JSON.stringify(payload);
    const headers = {
      Accept: "application/json",
      "Content-Type": "application/json",
    };

    if (bodyString) {
      headers["Content-Length"] = Buffer.byteLength(bodyString).toString();
    }

    return new Promise((resolve, reject) => {
      const req = transport.request(
        url,
        {
          method,
          headers,
        },
        (res) => {
          let raw = "";
          res.setEncoding("utf8");
          res.on("data", (chunk) => {
            raw += chunk;
          });
          res.on("end", () => {
            resolve([res.statusCode ?? 0, parseJsonOrText(raw)]);
          });
        },
      );

      req.setTimeout(this.timeoutMs, () => {
        req.destroy(new TestFailure(`Request timed out after ${this.timeoutMs}ms: ${method} ${url}`));
      });
      req.on("error", reject);

      if (bodyString) {
        req.write(bodyString);
      }
      req.end();
    });
  }

  printDebugRequest(method, url, payload) {
    const label = colorize("[DEBUG REQUEST]", Ansi.CYAN, this.color);
    console.log(`${label} ${method} ${url}`);
    console.log("Request JSON:");
    console.log(prettyJson(payload));
  }

  printDebugResponse(status, body) {
    const responseColor = status < 400 ? Ansi.GREEN : Ansi.RED;
    const label = colorize("[DEBUG RESPONSE]", responseColor, this.color);
    console.log(`${label} HTTP ${status}`);
    console.log("Response JSON:");
    console.log(prettyJson(body));
    console.log();
  }
}

function startServer(projectRoot) {
  const proc = spawn("cargo", ["run", "--quiet"], {
    cwd: projectRoot,
    stdio: ["ignore", "pipe", "pipe"],
  });

  let logs = "";
  proc.stdout.setEncoding("utf8");
  proc.stderr.setEncoding("utf8");
  proc.stdout.on("data", (chunk) => {
    logs += chunk;
  });
  proc.stderr.on("data", (chunk) => {
    logs += chunk;
  });

  return { proc, getLogs: () => logs.trim() };
}

async function stopServer(proc) {
  if (proc.exitCode !== null) {
    return;
  }

  proc.kill("SIGTERM");

  const exited = await Promise.race([
    new Promise((resolve) => {
      proc.once("exit", () => resolve(true));
    }),
    sleep(8_000).then(() => false),
  ]);

  if (!exited) {
    proc.kill("SIGKILL");
    await new Promise((resolve) => {
      proc.once("exit", () => resolve());
    });
  }
}

async function waitForServer(client, timeoutSeconds = 45, serverHandle = null) {
  const deadline = Date.now() + timeoutSeconds * 1000;

  while (Date.now() < deadline) {
    if (serverHandle && serverHandle.proc.exitCode !== null) {
      const output = serverHandle.getLogs();
      const details = output ? `\nServer output:\n${output}` : "";
      throw new TestFailure(`Backend process exited before becoming ready.${details}`);
    }

    try {
      const [status] = await client.requestJson(
        "GET",
        "/get-user",
        null,
        { email: "_probe_" },
        false,
      );
      if (status === 200) {
        return;
      }
    } catch {
      // Retry until timeout.
    }

    await sleep(1_000);
  }

  throw new TestFailure(`Server at ${client.baseUrl} was not reachable within ${timeoutSeconds}s`);
}

async function runSmokeTest(client) {
  const suffix = `${Math.floor(Date.now() / 1000)}_${rand(6)}`;
  const username = `test_user_${suffix}`;
  const email = `todo.${suffix}@example.com`;
  const password = `Pass_${suffix}!`;

  step("Create a new user");
  let [status, body] = await client.requestJson("POST", "/create-user", {
    username,
    email,
    password,
  });
  assert(status === 200, `/create-user expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body && typeof body === "object", `/create-user expected JSON object, got ${prettyJson(body)}`);
  assert(body.status === "Success", `/create-user failed: body=${prettyJson(body)}`);
  ok("User created");

  step("Fetch user by email");
  [status, body] = await client.requestJson("GET", "/get-user", null, { email });
  assert(status === 200, `/get-user expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.email === email, `/get-user email mismatch: body=${prettyJson(body)}`);
  assert(body?.username === username, `/get-user username mismatch: body=${prettyJson(body)}`);
  ok("User fetched");

  step("Authenticate with wrong password (negative test)");
  [status, body] = await client.requestJson("POST", "/auth", {
    email,
    pass: `${password}_wrong`,
  });
  assert(status === 200, `/auth (wrong pass) expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.auth === false, `/auth (wrong pass) auth should be false: body=${prettyJson(body)}`);
  assert(body?.error === true, `/auth (wrong pass) error should be true: body=${prettyJson(body)}`);
  assert(body?.jwt === null, `/auth (wrong pass) jwt should be null: body=${prettyJson(body)}`);
  ok("Wrong password rejected");

  step("Authenticate with valid password");
  [status, body] = await client.requestJson("POST", "/auth", {
    email,
    pass: password,
  });
  assert(status === 200, `/auth expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.auth === true, `/auth should return auth=true: body=${prettyJson(body)}`);
  assert(body?.error === false, `/auth should return error=false: body=${prettyJson(body)}`);
  const jwt = body?.jwt;
  assert(typeof jwt === "string" && jwt.length > 10, `/auth jwt missing/invalid: body=${prettyJson(body)}`);
  ok("Authentication succeeded");

  step("Verify valid JWT");
  [status, body] = await client.requestJson("POST", "/jwt-auth", { jwt });
  assert(status === 200, `/jwt-auth expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.tampered === false, `/jwt-auth tampered should be false: body=${prettyJson(body)}`);
  assert(body?.email === email, `/jwt-auth email mismatch: body=${prettyJson(body)}`);
  ok("Valid JWT accepted");

  step("Verify tampered JWT (negative test)");
  [status, body] = await client.requestJson("POST", "/jwt-auth", { jwt: `${jwt}.broken` });
  assert(status === 200, `/jwt-auth (tampered) expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.tampered === true, `/jwt-auth (tampered) should be true: body=${prettyJson(body)}`);
  assert(body?.email === null, `/jwt-auth (tampered) email should be null: body=${prettyJson(body)}`);
  ok("Tampered JWT rejected");

  step("Add two tasks");
  for (const [idx, name] of ["Write tests", "Ship release"].entries()) {
    [status, body] = await client.requestJson("POST", "/add-task", {
      name,
      priority: idx + 1,
      auth_jwt: jwt,
    });
    assert(status === 200, `/add-task expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
    assert(body === true, `/add-task expected true, got body=${prettyJson(body)}`);
  }
  ok("Tasks added");

  step("Get tasks with authenticated request");
  [status, body] = await client.requestJson("GET", "/get-tasks", {
    auth_jwt: jwt,
    counter: 10,
  });
  assert(status === 200, `/get-tasks expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.res === true, `/get-tasks should return res=true: body=${prettyJson(body)}`);
  let tasks = body?.tasks;
  assert(Array.isArray(tasks), `/get-tasks tasks should be a list: body=${prettyJson(body)}`);
  assert(tasks.length >= 1, `/get-tasks expected at least one task, got ${prettyJson(body)}`);
  const firstTask = tasks[0];
  assert("id" in firstTask && "name" in firstTask, `Unexpected task shape: ${prettyJson(firstTask)}`);
  ok("Tasks fetched");

  const firstTaskId = firstTask.id;
  const initialStatus = firstTask.status;
  assert(typeof initialStatus === "boolean", `Task status should be bool: ${prettyJson(firstTask)}`);

  step("Toggle first task status by id");
  [status, body] = await client.requestJson("POST", "/update-task", {
    id: firstTaskId,
    auth_jwt: jwt,
  });
  assert(status === 200, `/update-task expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.success === true, `/update-task expected success=true, got ${prettyJson(body)}`);
  ok("Task status toggled once");

  step("Verify first toggle changed task status");
  [status, body] = await client.requestJson("GET", "/get-tasks", {
    auth_jwt: jwt,
    counter: 10,
  });
  assert(status === 200, `/get-tasks (after 1st toggle) expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.res === true, `/get-tasks (after 1st toggle) should return res=true: body=${prettyJson(body)}`);
  tasks = body?.tasks;
  assert(Array.isArray(tasks), `/get-tasks (after 1st toggle) tasks should be list: body=${prettyJson(body)}`);
  const toggledTask = getTaskById(tasks, firstTaskId);
  assert(
    toggledTask.status === !initialStatus,
    `/update-task should toggle status from ${initialStatus} to ${!initialStatus}, got ${prettyJson(toggledTask)}`,
  );
  ok("First toggle verified");

  step("Toggle first task status back to original value");
  [status, body] = await client.requestJson("POST", "/update-task", {
    id: firstTaskId,
    auth_jwt: jwt,
  });
  assert(status === 200, `/update-task (2nd toggle) expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.success === true, `/update-task (2nd toggle) expected success=true, got ${prettyJson(body)}`);
  ok("Task status toggled twice");

  step("Verify second toggle restored original task status");
  [status, body] = await client.requestJson("GET", "/get-tasks", {
    auth_jwt: jwt,
    counter: 10,
  });
  assert(status === 200, `/get-tasks (after 2nd toggle) expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.res === true, `/get-tasks (after 2nd toggle) should return res=true: body=${prettyJson(body)}`);
  tasks = body?.tasks;
  assert(Array.isArray(tasks), `/get-tasks (after 2nd toggle) tasks should be list: body=${prettyJson(body)}`);
  const toggledBackTask = getTaskById(tasks, firstTaskId);
  assert(
    toggledBackTask.status === initialStatus,
    `/update-task second toggle should restore status=${initialStatus}, got ${prettyJson(toggledBackTask)}`,
  );
  ok("Second toggle verified");

  step("Update unknown task id (negative test)");
  [status, body] = await client.requestJson("POST", "/update-task", {
    id: firstTaskId + 10_000_000,
    auth_jwt: jwt,
  });
  assert(
    status === 500,
    `/update-task (unknown task id) expected HTTP 500, got ${status}, body=${prettyJson(body)}`,
  );
  assert(
    body?.success === false,
    `/update-task (unknown task id) should return success=false, got ${prettyJson(body)}`,
  );
  ok("Unknown task id rejected");

  step("Delete task with invalid JWT (negative test)");
  [status, body] = await client.requestJson("POST", "/delete-task", {
    auth_jwt: `${jwt}.broken`,
    task_id: firstTaskId,
  });
  assert(status === 500, `/delete-task (invalid jwt) expected HTTP 500, got ${status}, body=${prettyJson(body)}`);
  assert(body?.success === false, `/delete-task (invalid jwt) should fail: body=${prettyJson(body)}`);
  ok("Invalid-JWT delete rejected");

  step("Delete task with valid JWT");
  [status, body] = await client.requestJson("POST", "/delete-task", {
    auth_jwt: jwt,
    task_id: firstTaskId,
  });
  assert(status === 200, `/delete-task expected HTTP 200, got ${status}, body=${prettyJson(body)}`);
  assert(body?.success === true, `/delete-task should return success=true: body=${prettyJson(body)}`);
  ok("Task deleted");
}

function parseArgs(argv) {
  const args = {
    baseUrl: DEFAULT_BASE_URL,
    debug: false,
    startServer: false,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];

    if (arg === "-d" || arg === "--debug") {
      args.debug = true;
      continue;
    }

    if (arg === "--start-server") {
      args.startServer = true;
      continue;
    }

    if (arg === "--base-url") {
      const value = argv[i + 1];
      if (!value) {
        throw new TestFailure("Missing value for --base-url");
      }
      args.baseUrl = value;
      i += 1;
      continue;
    }

    if (arg === "-h" || arg === "--help") {
      printHelp();
      process.exit(0);
    }

    throw new TestFailure(`Unknown argument: ${arg}`);
  }

  return args;
}

function printHelp() {
  console.log(`Smoke test the Todo backend API

Usage:
  bun test/api_smoke_test.js [options]

Options:
  --base-url <url>    API base URL (default: ${DEFAULT_BASE_URL})
  -d, --debug         Print colorized request/response JSON for each API call
  --start-server      Start backend with cargo run --quiet before tests
  -h, --help          Show this help message
`);
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const client = new ApiClient(args.baseUrl, 10_000, args.debug, USE_COLOR);

  /** @type {{ proc: import("node:child_process").ChildProcessWithoutNullStreams, getLogs: () => string } | null} */
  let serverHandle = null;

  try {
    if (args.startServer) {
      console.log("[INFO] Starting backend server with cargo run --quiet");
      serverHandle = startServer(getProjectRoot());
    }

    console.log(`[INFO] Waiting for server: ${args.baseUrl}`);
    await waitForServer(client, 45, serverHandle);

    console.log("[INFO] Running smoke test flow");
    await runSmokeTest(client);

    console.log("\n[SUCCESS] All smoke tests passed");
    process.exitCode = 0;
  } catch (error) {
    if (error instanceof TestFailure) {
      console.error(`\n[FAIL] ${error.message}`);
    } else {
      const message = error instanceof Error ? error.message : String(error);
      console.error(`\n[ERROR] Unexpected exception: ${message}`);
    }
    process.exitCode = 1;
  } finally {
    if (serverHandle) {
      console.log("[INFO] Stopping backend server");
      await stopServer(serverHandle.proc);
    }
  }
}

await main();
